File: PRACTICAS/CAMBIAR_RUBY/RUBY2/Civitas/lib/tablero.rb
module Civitas
  class Tablero
    
    def initialize (num_casilla_carcel)

      if(num_casilla_carcel > 1)
        @num_casilla_carcel = num_casilla_carcel
      else
        @num_casilla_carcel = 1
      end
      salida = Casilla.casilla("Salida")
      @casillas = []
      @casillas << salida
      @por_salida =  0
      @tiene_juez = false

    end
    
    attr_accessor :num_casilla_carcel, :casillas, :por_salida, :tiene_juez
    
    def correcto

    return @casillas.size > @num_casilla_carcel && @tiene_juez
  end
    
    
    def get_por_salida
      if (@por_salida > 0)
        @por_salida = @por_salida -1
        return @por_salida +1
      else
        return @por_salida
      end
    end
    
    def aniade_casilla (casilla)
      carcel = Casilla.casilla("Carcel")

    if(@casillas.length == @num_casilla_carcel)

      @casillas << carcel
      @casillas << casilla


    else

      @casillas << casilla

      if(@casillas.length == @num_casilla_carcel)

      @casillas << carcel
      end


    end
    end
    
    def aniade_juez
      if (!@tiene_juez)
      juez = Casilla.new(@num_casilla_carcel, "Juez")
      aniade_casilla (juez)
      @tiene_juez = true

    end
    end
    
    def get_casilla (numCasilla)
      return @casillas[numCasilla]
    end
    
    def nueva_posicion (actual, tirada)
      if (correcto == false)
        return -1;
      else
        pos = (actual + tirada)%@casillas.size
        
      end
      return pos
    end
    
    def calcular_tirada (origen, destino)
      tirada = destino - origen
      if (tirada < 0)
        tirada = tirada + @casillas.length
      end
      return tirada
    end
    
    def borrar_casilla (ip)
      @casillas.delete_at(ip)
    end
    
    private :correcto
  end
end

File: PRACTICAS/PRACTICA_ACTUAL/JAVA/JAVA/Civitas_P1/src/civitas_p1/Tablero.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas_p1;
import java.util.*;

/**
 *
 * @author angelsolano
 */

public class Tablero {
    
    private int numCasillaCarcel; 
    private ArrayList<CasillaDescanso> casillas;
    private int porSalida;
    private boolean tieneJuez; 
    
    private boolean correcto(){
        return casillas.size() > numCasillaCarcel && tieneJuez;
    }
    
    private boolean correcto(int numCasilla){
        return correcto() && numCasilla >= 0 && numCasilla < casillas.size();
    }
    
    public Tablero(int numCasillaCarcel){
        
        if (numCasillaCarcel > 1)
            this.numCasillaCarcel = numCasillaCarcel;
        else
            this.numCasillaCarcel = 1;
        
        CasillaDescanso salida;
        salida = new CasillaDescanso("Salida");

        casillas = new ArrayList<CasillaDescanso>();
        casillas.add(salida);
        porSalida = 0;
        tieneJuez = false;
        
    }
    
    int getCarcel(){
        return numCasillaCarcel;
    }
    
    int getPorSalida(){
        
        if (porSalida > 0){
            porSalida --;
            return porSalida + 1;
        }
        else
            return porSalida;
     
    }
    
    void añadeCasilla(CasillaDescanso casilla){
        
        CasillaDescanso carcel;
        carcel = new CasillaDescanso("Carcel");
        
        if (casillas.size() == numCasillaCarcel){
            casillas.add(carcel);
            casillas.add(casilla);           
        }
        else{
            
            casillas.add(casilla);
            
            if (casillas.size() == numCasillaCarcel)
                casillas.add(carcel);
            
        }
            
    }
    
    void añadeJuez(){
        
        if (!tieneJuez){
            
            CasillaDescanso juez;
            juez = new CasillaJuez("Juez",numCasillaCarcel);
            
            
            añadeCasilla(juez);
            tieneJuez = true;
            
        }
    }
    
    CasillaDescanso getCasilla(int numCasilla){
        
        if (correcto(numCasilla))
            return casillas.get(numCasilla);
        else
            return null;
        
    }
    
    int nuevaPosicion(int actual, int tirada){
        
        if(!correcto())
            return -1;
        else{
            
            int nueva_pos;
            
            nueva_pos = (actual + tirada) % casillas.size();
            
            if (nueva_pos != (actual + tirada))
                porSalida ++;
            
            return nueva_pos;
            
        }
                
    }
    
    int calcularTirada(int origen, int destino){
        
        int tirada = destino - origen;
        
        if (tirada < 0)
            return tirada + casillas.size();
        else
            return tirada;
        
        
    }
    
    public ArrayList<CasillaDescanso> getCasillas(){
        return casillas;
    }
    
    
    
}


